File: gateway/app/services/ai/schemas/pipeline.py
from __future__ import annotations

from enum import Enum
from typing import Any, Dict, List, Optional

from pydantic import BaseModel


class ConfirmationStatus(str, Enum):
    CONFIRMED = "CONFIRMED"
    DISMISSED = "DISMISSED"


class CriticalityLevel(str, Enum):
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"
    CRITICAL = "CRITICAL"


class AlertEvent(BaseModel):
    event_id: str
    lat: float
    lon: float
    camera_id: Optional[str] = None
    image_url: Optional[str] = None
    timestamp: str


class CameraResult(BaseModel):
    confidence: float               # 0.0 – 1.0  fire/smoke detection confidence
    detected: bool
    image_url: Optional[str] = None
    raw: Optional[Dict[str, Any]] = None
    latency_ms: Optional[float] = None  # wall time for this agent (for SLO / debugging)
    telemetry: Optional[Dict[str, Any]] = None


class SatelliteResult(BaseModel):
    thermal_confidence: float       # 0.0 – 1.0
    hotspot_detected: bool
    raw: Optional[Dict[str, Any]] = None
    latency_ms: Optional[float] = None
    telemetry: Optional[Dict[str, Any]] = None


class WeatherResult(BaseModel):
    wind_speed: float               # m/s
    wind_direction: float           # degrees
    humidity: float                 # %
    spread_risk: float              # 0.0 – 1.0 calculated heuristic
    raw: Optional[Dict[str, Any]] = None
    latency_ms: Optional[float] = None
    telemetry: Optional[Dict[str, Any]] = None


class FusionResult(BaseModel):
    status: ConfirmationStatus
    combined_score: float
    reason: str
    telemetry: Optional[Dict[str, Any]] = None


class ReasoningResult(BaseModel):
    scene_description: str
    key_observations: List[str]


class ClassificationResult(BaseModel):
    criticality: CriticalityLevel
    score: float
    reasoning: str


class SuggestionResult(BaseModel):
    action_plan: List[str]
    alert_message: str
    recommended_resources: List[str]


class OutputResult(BaseModel):
    notification_sent: bool
    dashboard_updated: bool
    incident_id: str
    logged: bool


class PipelineResult(BaseModel):
    event_id: str
    camera: Optional[CameraResult] = None
    satellite: Optional[SatelliteResult] = None
    weather: Optional[WeatherResult] = None
    fusion: Optional[FusionResult] = None
    reasoning: Optional[ReasoningResult] = None
    classification: Optional[ClassificationResult] = None
    suggestion: Optional[SuggestionResult] = None
    output: Optional[OutputResult] = None
    error: Optional[str] = None

File: frontend/src/types/pipeline.ts
export type ConfirmationStatus = 'CONFIRMED' | 'DISMISSED'
export type CriticalityLevel = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL'

export interface AlertEvent {
  event_id: string
  lat: number
  lon: number
  camera_id?: string
  image_url?: string
  timestamp: string
}

export interface CameraResult {
  confidence: number
  detected: boolean
  image_url?: string | null
  raw?: Record<string, unknown> | null
  latency_ms?: number | null
  telemetry?: Record<string, unknown> | null
}

export interface SatelliteResult {
  thermal_confidence: number
  hotspot_detected: boolean
  raw?: Record<string, unknown> | null
  latency_ms?: number | null
  telemetry?: Record<string, unknown> | null
}

export interface WeatherResult {
  wind_speed: number
  wind_direction: number
  humidity: number
  spread_risk: number
  raw?: Record<string, unknown> | null
  latency_ms?: number | null
  telemetry?: Record<string, unknown> | null
}

export interface FusionResult {
  status: ConfirmationStatus
  combined_score: number
  reason: string
  telemetry?: Record<string, unknown> | null
}

export interface ReasoningResult {
  scene_description: string
  key_observations: string[]
}

export interface ClassificationResult {
  criticality: CriticalityLevel
  score: number
  reasoning: string
}

export interface SuggestionResult {
  action_plan: string[]
  alert_message: string
  recommended_resources: string[]
}

export interface OutputResult {
  notification_sent: boolean
  dashboard_updated: boolean
  incident_id: string
  logged: boolean
}

export interface PipelineResult {
  event_id: string
  camera?: CameraResult | null
  satellite?: SatelliteResult | null
  weather?: WeatherResult | null
  fusion?: FusionResult | null
  reasoning?: ReasoningResult | null
  classification?: ClassificationResult | null
  suggestion?: SuggestionResult | null
  output?: OutputResult | null
  error?: string | null
}

export interface CameraNode {
  id: string
  name: string
  code: string
  lat: number
  lon: number
  online: boolean
  alertcaId?: string
}
